src/GazeboUnderwater.cpp

```diff
@@ -197,4 +229,137 @@ namespace gazebo_underwater
+        math::Vector3 gzTorque = modelQuaternion.RotateVector(link->GetWorldTorque());
+        base::Vector6d forces;
+        forces << gzForce[0], gzForce[1], gzForce[2], gzTorque[0], gzTorque[1], gzTorque[2];
+        base::Vector6d compForce = (gzInertia * inverseInertia - base::Matrix6d::Identity()) * forces;
+
+        math::Vector3 linearComp(compForce[0], compForce[1], compForce[2]);
+        math::Vector3 angularComp(compForce[3], compForce[4], compForce[5]);
+
+
+        link->AddForceAtWorldPosition(modelQuaternion.RotateVectorReverse(linearComp),cogPosition);
+        link->AddTorque( modelQuaternion.RotateVectorReverse(angularComp) );
+    }
+
+    std::vector<base::Matrix6d> GazeboUnderwater::convertToMatrices(const std::string &matrices)
+    {
+        std::vector<base::Matrix6d> ret;
+        std::vector<std::string> splitted;
+        boost::algorithm::split_regex( splitted, matrices, boost::regex( "\n\n" ));
+        if (splitted.size() != 2 && splitted.size() != 6)
+            gzthrow("GazeboUnderwater: Damping Parameters has not 2 or 6 matrices!");
+        for( size_t i=0; i<splitted.size(); i++)
+            ret.push_back(convertToMatrix(splitted.at(i)));
+        return ret;
+    }
+
+    base::Matrix6d GazeboUnderwater::convertToMatrix(const std::string &matrix)
+    {
+        base::Matrix6d ret;
+        std::vector<std::string> splitted;
+        boost::split( splitted, matrix, boost::is_any_of( "\n" ), boost::token_compress_on );
+        if (splitted.size() != 6)
+            gzthrow("GazeboUnderwater: Matrix has not 6 lines!");
+        for( size_t i=0; i<6; i++)
+        {
+            std::vector<std::string> line;
+            boost::split( line, splitted[i], boost::is_any_of( " " ), boost::token_compress_on );
+            if (line.size() != 6)
+                gzthrow("GazeboUnderwater: Line has not 6 columns!");
+            for(size_t j=0; j<6; j++)
+                ret(i,j) = atof(line.at(j).c_str());
+        }
+        return ret;
+    }
+
+    base::Vector6d GazeboUnderwater::getModelFrameVelocities()
+    {
+        // Calculates the difference between the model and fluid velocity relative to the world frame
+        math::Vector3 velocityDifference = link->GetWorldLinearVel() - fluidVelocity;
+        math::Vector3 angularVelocity = link->GetWorldAngularVel();
+
+        math::Quaternion modelQuaternion = link->GetWorldCoGPose().rot;
+        math::Vector3 mfVelocity = modelQuaternion.RotateVector( velocityDifference );
+        math::Vector3 mfAngularVelocity = modelQuaternion.RotateVector( angularVelocity );
+        base::Vector6d velocities = base::Vector6d::Zero();
+        velocities << mfVelocity[0], mfVelocity[1], mfVelocity[2], mfAngularVelocity[0], mfAngularVelocity[1], mfAngularVelocity[2];
+        return velocities;
+    }
+
+    base::Matrix6d GazeboUnderwater::mountGzInertiaMatrix(ModelPtr model) const
+    {
+        base::Matrix6d gzInertia = base::Matrix6d::Zero();
```

Comment:
This should really be using the Gazebo representations ... one can get the inertia information per-link and add them up.